Tasks/NpmAuthenticate/npmauth.ts

```diff
@@ -0,0 +1,136 @@
+    let indexFile = saveNpmrcPath+'\\index.json';
+
+    if (fs.existsSync(indexFile)) { //If the file exists, add to it.
+        npmrcTable = JSON.parse(fs.readFileSync(indexFile, 'utf8'));
+        if (npmrcTable[npmrc]=== undefined) {
+            npmrcTable[npmrc] = npmrcTable['index'];
+            npmrcTable['index']++;
+            fs.writeFileSync(indexFile, JSON.stringify(npmrcTable));
+            util.saveFileWithName(npmrc, npmrcTable[npmrc], saveNpmrcPath);
+        }
+    }
+    else { //If the file doesn't exist, create it. 
+        npmrcTable = new Object();
+        npmrcTable['index']=0;
+        npmrcTable[npmrc] = npmrcTable['index'];
+        npmrcTable['index']++;
+        fs.writeFileSync(indexFile, JSON.stringify(npmrcTable));
+        util.saveFileWithName(npmrc, '0', saveNpmrcPath);
+
+    }
+
+    let endpointRegistries;
+    let endpointIds = tl.getDelimitedInput(constants_1.NpmAuthenticateTaskInput.CustomEndpoint, ',');
+    if (endpointIds && endpointIds.length > 0) {
+        endpointRegistries = endpointIds.map(e => npmregistry_1.NpmRegistry.FromServiceEndpoint(e, true));
+    }
+    let LocalNpmRegistries = await util.getLocalNpmRegistries(workingDir);
+    
+    
+    let npmrcFile = fs.readFileSync(npmrc, 'utf8').split('\n');
+    for (let fileLine of npmrcFile) {
+        //For each registry in the .npmrc file (every line with "registry=")
+        //If it's got a service endpoint, use that. If not, use a local registry.
+        //If it's got one of those, add the credentials into the .npmrc file
+        if (fileLine.search('registry=') > -1) {  
+            let keyAndValue = fileLine.split('=');
+            let registryURL = URL.parse(keyAndValue[1]);
+            let registry;
+            if (endpointRegistries && endpointRegistries.length > 0) {
+                for (let serviceEndpoint of endpointRegistries) {
+                    let serviceURL = URL.parse(serviceEndpoint.url);
+                    if (serviceURL.hostname == registryURL.hostname) {
+                        console.log(tl.loc("AddingEndpointCredentials", registryURL.hostname));
+                        registry = serviceEndpoint;
+                        npmrcFile = clearFileOfReferences(npmrc, npmrcFile, serviceURL);
+                        break;
+                    }
+                }
+            }
+            if (!registry) {
+                for (let localRegistry of LocalNpmRegistries) {
+                    let localURL = URL.parse(localRegistry.url);
+                    if (localURL.hostname == registryURL.hostname) {
+                        console.log(tl.loc("AddingLocalCredentials"))
+                        registry = localRegistry;
+                        npmrcFile = clearFileOfReferences(npmrc, npmrcFile, localURL);
+                        break;
+                    }
+                }
+            }
+            if (registry) {
+                if (registry.authOnly === false) {
```

Comment:
I thought we were never overriding registries in the .npmrc. If so, authOnly should only be true, meaning this block will never execute in this task. Am I missing something?